#### Project: CropGuard AI

The project uses two supervised ML tasks.

1. 🌾 Pre-Cultivation Crop Yield Prediction
ML Task: Regression
Target (Y): yield — existing numerical column
Purpose: Predict the expected crop yield in tonnes/hectare for a selected district, crop, season, and agricultural year using historical crop performance, rainfall, irrigation, and contextual information.
Output: Predicted Yield
Example: 4.82 tonnes/hectare
2. ⚠️ Agricultural Risk Classification
ML Task: Classification
Target (Y): risk_level — derived target
Classes: Low Risk / Moderate Risk / High Risk
Purpose: Identify the agricultural risk associated with a particular district–crop–season combination using historical crop performance, rainfall conditions, irrigation conditions, and other validated features.
Output: Risk Level + Risk Probability
Example: HIGH RISK — 78% probability

## Coding 

In [1]:
## importing libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import os 
path = r"C:\Users\SVS\Desktop\ML Project\data\ml_ready\Final_dataset.csv"
print(os.path.exists(path))

True


In [3]:
df = pd.read_csv(path)
print("Shape:", df.shape)


Shape: (10708, 29)


### 📤 0. Data Loading (from db load tables data as pandas df)


In [4]:
%pip install pymysql

Note: you may need to restart the kernel to use updated packages.


In [5]:
from sqlalchemy import create_engine
from urllib.parse import quote_plus

username = "root"
password = input("Enter MySQL password: ")

engine = create_engine(
    f"mysql+pymysql://{username}:{quote_plus(password)}@localhost:3306/cropguard_ai"
)

with engine.connect() as connection:
    print("Database connection successful!")

Enter MySQL password:  Raju8008


Database connection successful!


In [6]:
import os 
path = r"C:\Users\SVS\Desktop\ML Project\data\ml_ready\Final_dataset.csv"
print(os.path.exists(path))

True


In [7]:
df.to_sql(
    "validated_agriculture_data",
    con=engine,
    if_exists="replace",
    index=False
)

print("Data imported successfully!")

Data imported successfully!


check = pd.read_sql(
    "SELECT COUNT(*) AS total_rows FROM validated_agriculture_data",
    engine
)

print(check)

In [8]:
df = pd.read_sql(
    """
    SELECT *
    FROM validated_agriculture_data
    """,
    engine
)

print("Data loaded from MySQL successfully!")
print("Shape:", df.shape)
df.head()

Data loaded from MySQL successfully!
Shape: (10708, 29)


,id,year,state_name,state_code,district_name,district_code,crop_name,crop_code,crop_type,season,...,normal_rainfall,rainfall_deviation,agri_year,irrigation_food_category,irrigation_crop_type,irrigation_crop_name,irrigation_crop_code,irrigation_season,crop_irrigation_area,total_irrigated_area
0,385411,2019-2020,Telangana,36,Adilabad,501,Bajra,103.0,Cereals,Kharif,...,1103.8,69.931180,2019-2020,Food Crop,Cereals & Millets,Bajra,103.0,Kharif,79570.0,1048794.0
1,385812,2019-2020,Telangana,36,Kumuram Bheem Asifabad,699,Bajra,103.0,Cereals,Kharif,...,1103.8,69.931180,2019-2020,Food Crop,Cereals & Millets,Bajra,103.0,Kharif,79570.0,1048794.0
2,407355,2020-2021,Telangana,36,Kumuram Bheem Asifabad,699,Bajra,103.0,Cereals,Kharif,...,1103.6,-15.100530,2020-2021,Food Crop,Cereals & Millets,Bajra,103.0,Kharif,79635.0,1101556.0
3,450641,2022-2023,Telangana,36,Kumuram Bheem Asifabad,699,Bajra,103.0,Cereals,Kharif,...,1103.6,68.352253,2022-2023,Food Crop,Cereals & Millets,Bajra,103.0,Kharif,79128.0,1040598.0
4,284709,2014-2015,Telangana,36,Adilabad,501,Gram,201.0,Pulses,Rabi,...,1103.6,89.294537,2014-2015,Food Crop,Pulses,Gram,201.0,Rabi,2019.0,271774.0


## REGRESSION TASK — CROP YIELD PREDICTION

## 1. Modeling 

- ## 1.1 Pre-Requisites

In [9]:
## y = df["yield"]

#### 1.1.1 X & y:
- Picking y output col for Predictive Modelling & considering remaining cols are X

In [10]:
## Target (y)
y = df["yield"]

In [11]:
## Features (X)
X = df[
    [
        "district_name",
        "crop_name",
        "crop_type",
        "season",
        "area",
        "actual_rainfall",
        "normal_rainfall",
        "rainfall_deviation",
        "total_irrigated_area"
    ]
]

#### 1.1.2 Train-Test Split:¶
Dividing X & y data into train-test parts , where train part will be used for model learning , test part will be used for trained model testing
Thumb rules: 75-25, 80-20, 70-30

Since this project is intended for pre-cultivation prediction, the data is
split chronologically rather than randomly.
Older agricultural years are used for model training and the most recent
years are kept for testing. This gives a more realistic estimate of how the
model may perform on future agricultural seasons.

In [12]:
## starting year

df["year_start"] = (
    df["agri_year"]
    .astype(str)
    .str.extract(r"(\d{4})")[0]
    .astype(int)
)

In [13]:
df = df.sort_values(
    "year_start"
).reset_index(drop=True)
years = sorted(df["year_start"].unique())
print("Available agricultural years:")
print(years)

Available agricultural years:
[np.int64(2014), np.int64(2015), np.int64(2016), np.int64(2017), np.int64(2018), np.int64(2019), np.int64(2020), np.int64(2021), np.int64(2022)]


In [14]:
# CLEAN CATEGORICAL COLUMNS
for col in [
    "district_name",
    "crop_name",
    "crop_type",
    "season"
]:

    df[col] = (
        df[col]
        .astype(str)
        .str.strip()
        .str.lower()
    )


In [15]:
#REMOVE MISSING TARGET VALUES

df = df.dropna( subset=["yield"]).copy()


In [16]:
#  FEATURES

feature_columns = [

    "district_name",
    "crop_name",
    "crop_type",
    "season",

    "actual_rainfall",
    "normal_rainfall",
    "rainfall_deviation",
    "total_irrigated_area"

]

X = df[feature_columns].copy()
y = df["yield"].copy()


In [17]:
#  CHRONOLOGICAL TRAIN-TEST SPLIT

years = sorted(
    df["year_start"].unique()
)

split_index = int(
    len(years) * 0.80
)

train_years = years[:split_index]

test_years = years[split_index:]


print("Training years:")
print(train_years)

print("\nTesting years:")
print(test_years)


train_mask = df["year_start"].isin(
    train_years
)

test_mask = df["year_start"].isin(
    test_years
)


X_train = X.loc[train_mask].copy()

X_test = X.loc[test_mask].copy()

y_train = y.loc[train_mask].copy()

y_test = y.loc[test_mask].copy()


print("\nX_train shape:", X_train.shape)

print("X_test shape :", X_test.shape)

print("y_train shape:", y_train.shape)

print("y_test shape :", y_test.shape)


Training years:
[np.int64(2014), np.int64(2015), np.int64(2016), np.int64(2017), np.int64(2018), np.int64(2019), np.int64(2020)]

Testing years:
[np.int64(2021), np.int64(2022)]

X_train shape: (7212, 8)
X_test shape : (3496, 8)
y_train shape: (7212,)
y_test shape : (3496,)


####  1.1.3 Missing Values & Outliers Handling

Missing values are checked separately for input and target variables.

Missing values in the target are removed because the model cannot learn from
a record without a known yield.

Input missing values will be handled later through the preprocessing pipeline.

Yield outliers are investigated rather than automatically removed because
extreme values may represent genuine agricultural differences between crops.

In [18]:
print("Missing values in X_train:")
print(X_train.isnull().sum())

Missing values in X_train:
district_name           0
crop_name               0
crop_type               0
season                  0
actual_rainfall         0
normal_rainfall         0
rainfall_deviation      0
total_irrigated_area    0
dtype: int64


In [19]:
print("Missing values in X_test:")
print(X_test.isnull().sum())

Missing values in X_test:
district_name           0
crop_name               0
crop_type               0
season                  0
actual_rainfall         0
normal_rainfall         0
rainfall_deviation      0
total_irrigated_area    0
dtype: int64


In [20]:
## outliers Identification and Handling 
print("Yield statistics:")
print(y_train.describe())

Yield statistics:
count     7212.000000
mean        11.787729
std        351.602124
min          0.000000
25%          0.838000
50%          1.636000
75%          3.826500
max      21105.000000
Name: yield, dtype: float64


In [21]:
## IQR Method 
Q1 = y_train.quantile(0.25)
Q3 = y_train.quantile(0.75)

IQR = Q3 - Q1

lower_limit = Q1 - 1.5 * IQR
upper_limit = Q3 + 1.5 * IQR
print("Lower limit:", lower_limit)
print("Upper limit:", upper_limit)

Lower limit: -3.64475
Upper limit: 8.30925


In [22]:
## potential outliers 
outlier_mask = (
    (y_train < lower_limit) |
    (y_train > upper_limit)
)

print("Potential yield outliers:", outlier_mask.sum())

Potential yield outliers: 809


In [23]:
## Inspecting outliers 
outlier_rows = X_train.loc[outlier_mask].copy()

outlier_rows["yield"] = y_train.loc[outlier_mask]

display(
    outlier_rows[
        [
            "district_name",
            "crop_name",
            "crop_type",
            "season",
            "actual_rainfall",
            "normal_rainfall",
            "rainfall_deviation",
            "total_irrigated_area",
            "yield"
        ]
    ].sort_values("yield", ascending=False).head(20)
)

,district_name,crop_name,crop_type,season,actual_rainfall,normal_rainfall,rainfall_deviation,total_irrigated_area,yield
146,ranga reddy,coconut,oilseeds,whole year,779.570,835.40,78.692111,151208.0,21105.000
106,khammam,coconut,oilseeds,whole year,935.570,1095.70,6.442842,703084.0,21105.000
6020,khammam,sugarcane,sugar,kharif,1779.750,1095.70,37.852206,1168520.0,112.601
6167,wanaparthy,sugarcane,sugar,kharif,1118.360,749.00,73.800665,1903948.0,99.117
6812,yadadri bhuvanagiri,sugarcane,sugar,kharif,1123.495,744.30,40.205915,2139835.0,98.000
6767,ranga reddy,sugarcane,sugar,kharif,1131.060,835.40,32.711099,379501.0,98.000
5753,adilabad,sugarcane,sugar,kharif,1114.720,1103.60,-15.100530,1101556.0,98.000
6026,mahabubabad,sugarcane,sugar,kharif,1772.085,990.60,49.970233,1759290.0,98.000
5746,nirmal,sugarcane,sugar,kharif,1114.720,1103.60,-15.100530,1101556.0,98.000
5752,kumuram bheem asifabad,sugarcane,sugar,kharif,1114.720,1103.60,-15.100530,1101556.0,98.000


In [24]:
## outliers in Numerical X Features 
numerical_cols = [
    "actual_rainfall",
    "normal_rainfall",
    "rainfall_deviation",
    "total_irrigated_area"
]

In [25]:
for col in numerical_cols:
    Q1 = X_train[col].quantile(0.25)
    Q3 = X_train[col].quantile(0.75)
    IQR = Q3 - Q1

    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR

    count = (
        (X_train[col] < lower) |
        (X_train[col] > upper)
    ).sum()

    print(f"{col}: {count} potential outliers")

actual_rainfall: 0 potential outliers
normal_rainfall: 0 potential outliers
rainfall_deviation: 49 potential outliers
total_irrigated_area: 0 potential outliers


### Outlier Decision

The IQR method identified potential outliers in yield, area and rainfall
deviation.

The identified observations are retained because extreme values can occur
naturally in agricultural data. In particular, yield and cultivated area can
vary substantially across different crops and districts.

No observations are removed only on the basis of the IQR rule. Model
performance will be evaluated with the original observations, and the effect
of extreme values will be considered during model evaluation.

#### 1.1.4 Feature Engineering of x for y

#### 1.1.4.1 Feature Generation

For the baseline regression model, the available dataset already contains
the required agricultural, rainfall and irrigation features.

Therefore, no additional feature generation is performed at this stage.
The selected features will be evaluated directly during model training.

#### 1.1.4.2 Feature Selection
The following features were selected based on their relevance to crop yield
prediction and their availability in the prepared dataset.

In [26]:
selected_features = [
    "district_name",
    "crop_name",
    "crop_type",
    "season",
    "actual_rainfall",
    "normal_rainfall",
    "rainfall_deviation",
    "total_irrigated_area"
]

print("Selected features:")
for feature in selected_features:
 print(feature)

Selected features:
district_name
crop_name
crop_type
season
actual_rainfall
normal_rainfall
rainfall_deviation
total_irrigated_area


#### 1.1.4.3 Feature Type Identification

The selected input variables contain both categorical and numerical data.

Categorical variables will be encoded into numerical form because machine
learning algorithms require numerical input.

Numerical variables will be considered for scaling, especially for
distance-based algorithms such as KNN and SVR.

In [27]:
categorical_features = [
    "district_name",
    "crop_name",
    "crop_type",
    "season"
]

numerical_features = [
    "actual_rainfall",
    "normal_rainfall",
    "rainfall_deviation",
    "total_irrigated_area"
]

print("Categorical features:")
print(categorical_features)

print("\nNumerical features:")
print(numerical_features)

Categorical features:
['district_name', 'crop_name', 'crop_type', 'season']

Numerical features:
['actual_rainfall', 'normal_rainfall', 'rainfall_deviation', 'total_irrigated_area']


#### 1.1.4.4 Missing Value Handling

Missing numerical values will be replaced using the median calculated from
the training data.

Missing categorical values will be replaced using the most frequent value
from the training data.

This prevents information from the test dataset from influencing the
training process.

#### 1.1.4.5 Encoding 
one-hot encoding for categorical columns 

In [28]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

categorical_transformer = Pipeline(
    steps=[
        ("imputer", SimpleImputer(strategy="most_frequent")),
        ("encoder", OneHotEncoder(handle_unknown="ignore"))
    ]
)

#### 1.1.4.6 Scaling
For numerical features

In [29]:
from sklearn.preprocessing import StandardScaler
numerical_transformer = Pipeline(
    steps=[
        ("imputer", SimpleImputer(strategy="median")),
        ("scaler", StandardScaler())
    ]
)

In [30]:
preprocessor = ColumnTransformer(
    transformers=[
        ("num", numerical_transformer, numerical_features),
        ("cat", categorical_transformer, categorical_features)
    ]
)

print("Preprocessing pipeline created successfully.")

Preprocessing pipeline created successfully.


In [31]:
print(preprocessor)

ColumnTransformer(transformers=[('num',
                                 Pipeline(steps=[('imputer',
                                                  SimpleImputer(strategy='median')),
                                                 ('scaler', StandardScaler())]),
                                 ['actual_rainfall', 'normal_rainfall',
                                  'rainfall_deviation',
                                  'total_irrigated_area']),
                                ('cat',
                                 Pipeline(steps=[('imputer',
                                                  SimpleImputer(strategy='most_frequent')),
                                                 ('encoder',
                                                  OneHotEncoder(handle_unknown='ignore'))]),
                                 ['district_name', 'crop_name', 'crop_type',
                                  'season'])])


### 1.2 Model + Define, Train & Study

#### 1.2.1 Import Regression Algorithms

Different regression algorithms are trained and compared to identify the
model that performs best for crop yield prediction.

The following algorithms are considered:

1. Linear Regression
2. Polynomial Regression
3. Lasso Regression
4. Ridge Regression
5. KNN Regression
6. Support Vector Regression
7. Decision Tree Regression
8. Random Forest Regression

In [32]:
## impoerting models
from sklearn.linear_model import LinearRegression, Lasso, Ridge
from sklearn.preprocessing import PolynomialFeatures
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import SVR
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor

#### 1.2.2 Define Regression Models

In [33]:
models = {
    "Linear Regression": Pipeline([
        ("preprocessor", preprocessor),
        ("model", LinearRegression())
    ]),

    "Polynomial Regression": Pipeline([
        ("preprocessor", preprocessor),
        ("poly", PolynomialFeatures(degree=2, include_bias=False)),
        ("model", LinearRegression())
    ]),

    "Lasso Regression": Pipeline([
        ("preprocessor", preprocessor),
        ("model", Lasso(alpha=0.01, max_iter=10000))
    ]),

    "Ridge Regression": Pipeline([
        ("preprocessor", preprocessor),
        ("model", Ridge(alpha=1.0))
    ]),

    "KNN Regression": Pipeline([
        ("preprocessor", preprocessor),
        ("model", KNeighborsRegressor(n_neighbors=5))
    ]),

    "SVR": Pipeline([
        ("preprocessor", preprocessor),
        ("model", SVR())
    ]),

    "Decision Tree": Pipeline([
        ("preprocessor", preprocessor),
        ("model", DecisionTreeRegressor(random_state=42))
    ]),

    "Random Forest": Pipeline([
        ("preprocessor", preprocessor),
        ("model", RandomForestRegressor(
            n_estimators=100,
            random_state=42,
            n_jobs=-1
        ))
    ])
}

In [34]:
print("Regression models selected:\n")
for name in models:
    print( name)

Regression models selected:

Linear Regression
Polynomial Regression
Lasso Regression
Ridge Regression
KNN Regression
SVR
Decision Tree
Random Forest


#### 1.2.4 Train Models

In [35]:
trained_models = {}

for name, model in models.items():

    print(f"Training {name}...")

    model.fit(X_train, y_train)

    trained_models[name] = model

    print(f"{name} training completed.")

Training Linear Regression...
Linear Regression training completed.
Training Polynomial Regression...
Polynomial Regression training completed.
Training Lasso Regression...
Lasso Regression training completed.
Training Ridge Regression...
Ridge Regression training completed.
Training KNN Regression...
KNN Regression training completed.
Training SVR...
SVR training completed.
Training Decision Tree...
Decision Tree training completed.
Training Random Forest...
Random Forest training completed.


### 1.3 Trained Model Predictions & Evaluations

#### 1.3.1 Predictions on Train and Test Data

Predictions are generated using all trained regression models on both
training and test datasets.

The training predictions are used to understand model fitting, while
test predictions are used to evaluate how well the model performs on
unseen data.

In [36]:
predictions = {}

for name, model in trained_models.items():

    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)

    predictions[name] = {
        "y_train_pred": y_train_pred,
        "y_test_pred": y_test_pred
    }

    print(f"{name} predictions completed.")

Linear Regression predictions completed.
Polynomial Regression predictions completed.
Lasso Regression predictions completed.
Ridge Regression predictions completed.
KNN Regression predictions completed.
SVR predictions completed.
Decision Tree predictions completed.
Random Forest predictions completed.


#### 1.3.2 Regression Evaluation Metrics

In [37]:
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

results = []
for name in predictions:

    y_train_pred = predictions[name]["y_train_pred"]
    y_test_pred = predictions[name]["y_test_pred"]

    train_rmse = np.sqrt(mean_squared_error(y_train, y_train_pred))
    test_rmse = np.sqrt(mean_squared_error(y_test, y_test_pred))

    train_r2 = r2_score(y_train, y_train_pred)
    test_r2 = r2_score(y_test, y_test_pred)

    results.append({
        "Models": name,
        "TrainLoss (RMSE)": train_rmse,
        "TestLoss (RMSE)": test_rmse,
        "TrainScore (R²)": train_r2,
        "TestScore (R²)": test_r2
    })

#### 1.3.3 Bias-Variance Trade-off

In [38]:
for result in results:

    train_rmse = result["TrainLoss (RMSE)"]
    test_rmse = result["TestLoss (RMSE)"]

    train_r2 = result["TrainScore (R²)"]
    test_r2 = result["TestScore (R²)"]

    rmse_gap = test_rmse - train_rmse
    r2_gap = train_r2 - test_r2

    if train_r2 < 0.50 and test_r2 < 0.50:
        fit = "Underfit"

    elif r2_gap > 0.15:
        fit = "Overfit"

    else:
        fit = "Goodfit"

    result["Bias-Variance (Fit)"] = fit

#### 1.3.4 Final Regression Evaluation Table

In [39]:
evaluation_table = pd.DataFrame(results)

evaluation_table[
    [
        "Models",
        "TrainLoss (RMSE)",
        "TestLoss (RMSE)",
        "TrainScore (R²)",
        "TestScore (R²)",
        "Bias-Variance (Fit)"
    ]
]

,Models,TrainLoss (RMSE),TestLoss (RMSE),TrainScore (R²),TestScore (R²),Bias-Variance (Fit)
0,Linear Regression,3.036931e+00,3.940847,0.999925,0.903613,Goodfit
1,Polynomial Regression,2.154126e+00,12.336619,0.999962,0.055436,Overfit
2,Lasso Regression,3.148613e+00,3.925836,0.999920,0.904346,Goodfit
3,Ridge Regression,1.129328e+02,14.829830,0.896820,-0.364934,Overfit
4,KNN Regression,2.811276e+02,4.147337,0.360612,0.893248,Goodfit
5,SVR,3.515349e+02,9.852990,0.000244,0.397475,Underfit
6,Decision Tree,1.137820e-17,3.662378,1.000000,0.916754,Goodfit
7,Random Forest,6.326544e+01,3.459589,0.967619,0.925717,Goodfit


In [40]:
evaluation_table = evaluation_table.round({
    "TrainLoss (RMSE)": 2,
    "TestLoss (RMSE)": 2,
    "TrainScore (R²)": 2,
    "TestScore (R²)": 2
})

evaluation_table

,Models,TrainLoss (RMSE),TestLoss (RMSE),TrainScore (R²),TestScore (R²),Bias-Variance (Fit)
0,Linear Regression,3.04,3.94,1.00,0.90,Goodfit
1,Polynomial Regression,2.15,12.34,1.00,0.06,Overfit
2,Lasso Regression,3.15,3.93,1.00,0.90,Goodfit
3,Ridge Regression,112.93,14.83,0.90,-0.36,Overfit
4,KNN Regression,281.13,4.15,0.36,0.89,Goodfit
5,SVR,351.53,9.85,0.00,0.40,Underfit
6,Decision Tree,0.00,3.66,1.00,0.92,Goodfit
7,Random Forest,63.27,3.46,0.97,0.93,Goodfit


#### 1.3.5 Pick the best Model

In [41]:
model_ranking = evaluation_table.sort_values(
    by="TestLoss (RMSE)",
    ascending=True
).reset_index(drop=True)

model_ranking

,Models,TrainLoss (RMSE),TestLoss (RMSE),TrainScore (R²),TestScore (R²),Bias-Variance (Fit)
0,Random Forest,63.27,3.46,0.97,0.93,Goodfit
1,Decision Tree,0.00,3.66,1.00,0.92,Goodfit
2,Lasso Regression,3.15,3.93,1.00,0.90,Goodfit
3,Linear Regression,3.04,3.94,1.00,0.90,Goodfit
4,KNN Regression,281.13,4.15,0.36,0.89,Goodfit
5,SVR,351.53,9.85,0.00,0.40,Underfit
6,Polynomial Regression,2.15,12.34,1.00,0.06,Overfit
7,Ridge Regression,112.93,14.83,0.90,-0.36,Overfit


In [42]:
best_model_name = model_ranking.loc[0, "Models"]
print("Best Regression Model:", best_model_name)

Best Regression Model: Random Forest


Based on the evaluation results, Random Forest was selected as the best regression model for crop yield prediction.

The model achieved the lowest Test RMSE of 3.46 and a Test R² of 0.93 among the evaluated models. It also showed a good balance between training and testing performance, indicating a good fit without a significant bias-variance problem.

Decision Tree and Lasso Regression also achieved strong testing performance, but Random Forest produced the lowest Test RMSE and the highest Test R², making it the best-performing model overall.

#### 1.4 Hyperparameter Tuning

Hyperparameter tuning is performed to determine whether the performance
of the selected Random Forest regression model can be improved further.

Since Random Forest achieved the best overall performance during model
comparison, its important hyperparameters are tuned to obtain a better
balance between training and testing performance and to reduce the risk
of overfitting.

The tuned Random Forest model is evaluated on the test data using RMSE
and R² and compared with the original Random Forest model.

The final model is selected based on the lowest Test RMSE and highest
Test R².

### 1.5 Saving Model & Real-Time Prediction

#### 1.5.1 Select Final Model

Based on the model evaluation , Random Forest Regression was selected as the final model for crop yield prediction.

The Random Forest model achieved the lowest Test RMSE of 3.46 and a Test R² of 0.93 among the evaluated regression models, indicating strong predictive performance and good generalization on unseen data.

The final Random Forest pipeline includes the required preprocessing steps for numerical and categorical features along with the tuned Random Forest regression model.

In [43]:
# Select the best trained model
final_model = trained_models[best_model_name]
print("Final model selected:", best_model_name)

Final model selected: Random Forest


In [44]:
import joblib
joblib.dump(final_model, "cropguard_yield_model.pkl")
print("Model saved successfully.")

Model saved successfully.


In [45]:
loaded_model = joblib.load("cropguard_yield_model.pkl")
print("Saved model loaded successfully.")

Saved model loaded successfully.


In [46]:
# REAL-TIME CROP YIELD PREDICTION

import ipywidgets as widgets
from IPython.display import display, clear_output


# CREATE YEAR COLUMN

if "year_start" not in df.columns:

    df["year_start"] = (
        df["agri_year"]
        .astype(str)
        .str.extract(r"(\d{4})")[0]
        .astype(int)
    )


# DISTRICT DROPDOWN

districts = sorted(
    df["district_name"]
    .dropna()
    .unique()
    .tolist()
)

district_dropdown = widgets.Dropdown(
    options=districts,
    description="District:",
    style={"description_width": "80px"},
    layout=widgets.Layout(width="480px")
)


# CROP DROPDOWN

crop_dropdown = widgets.Dropdown(
    options=[],
    description="Crop:",
    style={"description_width": "80px"},
    layout=widgets.Layout(width="480px")
)


# SEASON DROPDOWN

season_dropdown = widgets.Dropdown(
    options=[],
    description="Season:",
    style={"description_width": "80px"},
    layout=widgets.Layout(width="480px")
)


# FARMER AREA INPUT

area_input = widgets.FloatText(
    value=1.0,
    min=0.01,
    description="Area:",
    style={"description_width": "80px"},
    layout=widgets.Layout(width="480px")
)


# PREDICT BUTTON

predict_button = widgets.Button(
    description="Predict Yield",
    button_style="success",
    layout=widgets.Layout(
        width="260px",
        height="42px"
    )
)


# OUTPUT AREA

output = widgets.Output(
    layout=widgets.Layout(
        width="480px",
        border="1px solid lightgray",
        padding="10px"
    )
)


# UPDATE CROPS WHEN DISTRICT CHANGES

def update_crops(change):

    selected_district = change["new"]

    if selected_district is None:

        crop_dropdown.options = []
        season_dropdown.options = []

        return


    district_data = df[
        df["district_name"] == selected_district
    ]


    available_crops = sorted(
        district_data["crop_name"]
        .dropna()
        .unique()
        .tolist()
    )


    crop_dropdown.options = available_crops
    season_dropdown.options = []


    if available_crops:

        crop_dropdown.value = available_crops[0]

    else:

        crop_dropdown.value = None


# OBSERVE DISTRICT CHANGES

district_dropdown.observe(
    update_crops,
    names="value"
)


# UPDATE SEASONS WHEN CROP CHANGES

def update_seasons(change):

    selected_district = district_dropdown.value

    selected_crop = change["new"]


    if (
        selected_district is None
        or selected_crop is None
    ):

        season_dropdown.options = []

        return


    crop_data = df[
        (df["district_name"] == selected_district) &
        (df["crop_name"] == selected_crop)
    ]


    available_seasons = sorted(
        crop_data["season"]
        .dropna()
        .unique()
        .tolist()
    )


    season_dropdown.options = available_seasons


    if available_seasons:

        season_dropdown.value = available_seasons[0]

    else:

        season_dropdown.value = None


# OBSERVE CROP CHANGES

crop_dropdown.observe(
    update_seasons,
    names="value"
)


# REAL-TIME PREDICTION FUNCTION

def predict_yield(button):

    with output:

        clear_output()


        # GET USER INPUTS

        selected_district = district_dropdown.value

        selected_crop = crop_dropdown.value

        selected_season = season_dropdown.value

        farmer_area = area_input.value


        # VALIDATE AREA

        if farmer_area <= 0:

            print(
                "Please enter a valid area greater than 0 hectare."
            )

            return


        # VALIDATE SELECTION

        if (
            selected_district is None
            or selected_crop is None
            or selected_season is None
        ):

            print(
                "Please select District, Crop and Season."
            )

            return


        # FIND MATCHING HISTORICAL RECORDS

        matching_data = df[
            (df["district_name"] == selected_district) &
            (df["crop_name"] == selected_crop) &
            (df["season"] == selected_season)
        ].copy()


        # CHECK DATA AVAILABILITY

        if matching_data.empty:

            print(
                "No historical data is available "
                "for the selected District, Crop and Season."
            )

            return


        # SELECT LATEST AVAILABLE RECORD

        matching_data = matching_data.sort_values(
            "year_start"
        )

        selected_data = matching_data.iloc[-1]


        # GET BACKEND FEATURES

        crop_type = selected_data["crop_type"]

        actual_rainfall = selected_data["actual_rainfall"]

        normal_rainfall = selected_data["normal_rainfall"]

        rainfall_deviation = selected_data["rainfall_deviation"]

        total_irrigated_area = selected_data[
            "total_irrigated_area"
        ]


        # CREATE MODEL INPUT
        # Area is not used as a model feature.

        sample_input = pd.DataFrame({

            "district_name": [
                selected_district
            ],

            "crop_name": [
                selected_crop
            ],

            "crop_type": [
                crop_type
            ],

            "season": [
                selected_season
            ],

            "actual_rainfall": [
                actual_rainfall
            ],

            "normal_rainfall": [
                normal_rainfall
            ],

            "rainfall_deviation": [
                rainfall_deviation
            ],

            "total_irrigated_area": [
                total_irrigated_area
            ]

        })


        # PREDICT YIELD

        predicted_yield = loaded_model.predict(
            sample_input
        )[0]


        # SAFETY CHECK

        predicted_yield = max(
            float(predicted_yield),
            0
        )


        # CALCULATE TOTAL PRODUCTION

        estimated_total = (
            predicted_yield * farmer_area
        )


        # DISPLAY RESULT

        print("=" * 60)

        print(
            "             🌾 CROP YIELD PREDICTION"
        )

        print("=" * 60)

        print(
            f"District        : "
            f"{str(selected_district).title()}"
        )

        print(
            f"Crop            : "
            f"{str(selected_crop).title()}"
        )

        print(
            f"Season          : "
            f"{str(selected_season).title()}"
        )

        print(
            f"Farmer Area     : "
            f"{farmer_area:.2f} Hectare"
        )

        print("-" * 60)

        print(
            f"Predicted Yield : "
            f"{predicted_yield:.2f} Tonnes/Hectare"
        )

        print(
            f"Estimated Total : "
            f"{estimated_total:.2f} Tonnes"
        )

        print("=" * 60)


# CONNECT BUTTON

predict_button.on_click(
    predict_yield
)


# INITIALIZE DROPDOWNS

update_crops({
    "new": district_dropdown.value
})


# DISPLAY USER INTERFACE

display(
    district_dropdown,
    crop_dropdown,
    season_dropdown,
    area_input,
    predict_button,
    output
)

Dropdown(description='District:', layout=Layout(width='480px'), options=('adilabad', 'bhadradri kothagudem', '…

Dropdown(description='Crop:', layout=Layout(width='480px'), options=('arhar/tur', 'bajra', 'banana', 'cashewnu…

Dropdown(description='Season:', layout=Layout(width='480px'), options=('all seasons', 'kharif', 'rabi'), style…

FloatText(value=1.0, description='Area:', layout=Layout(width='480px'), style=DescriptionStyle(description_wid…

Button(button_style='success', description='Predict Yield', layout=Layout(height='42px', width='260px'), style…

Output(layout=Layout(border_bottom='1px solid lightgray', border_left='1px solid lightgray', border_right='1px…